doc/man/dumper.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH DUMPER "1" "July 2025" "dumper v6.0.0" "User Commands"
.SH NAME
dumper \- manual page for dumper v6.0.0
.SH SYNOPSIS
.B dumper
[\fI\,-t\/\fR] [\fI\,-H height\/\fR]
.SH DESCRIPTION
.TP
\fB\-h\fR, \fB\-\-help\fR:
print this help message
.HP
\fB\-t\fR, \fB\-\-testnet\fR: use testnet (default: mainnet)
.TP
\fB\-H\fR, \fB\-\-height\fR:
use given height (default: best known tip)
.SH COPYRIGHT

Copyright (C) 2009-2025 The Bitcoin Core Developers
Copyright (C) 2015-2025 The Zcash Developers
Copyright (C) 2015-2025 Zdeveloper.org
Copyright (C) 2015-2025 Zen Blockchain Foundation

This is experimental software.

Distributed under the MIT software license, see the accompanying file COPYING
or <http://www.opensource.org/licenses/mit-license.php>.

This product includes software developed by the OpenSSL Project for use in the
OpenSSL Toolkit <https://www.openssl.org/> and cryptographic software written
by Eric Young.


doc/release-notes.md
(note: this is a temporary file, to be added-to by anybody, and moved to
release-notes at release time)

Notable changes
===============

Removal of time adjustment
-------------------------------------------------------------

Prior to v2.0.21, `zend` would adjust the local time that it used by up
to 70 minutes, according to a median of the times sent by the first 200 peers
to connect to it. This mechanism was inherently insecure, since an adversary
making multiple connections to the node could effectively control its time
within that +/- 70 minute window (this is called a "timejacking attack").

In the v2.0.21 release, in addition to other mitigations for timejacking attacks,
as a simplification the time adjustment code has now been completely removed.
Node operators should instead simply ensure that local time is set
reasonably accurately.

If it appears that the node has a significantly different time than its peers,
a warning will still be logged and indicated on the metrics screen if enabled.


src/dumper.cpp
// Copyright (c) 2025 The Horizen Foundation
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <getopt.h>
#include <unordered_map>

#include "txdb.h"
#include "chainparams.h"
#include "script/script.h"
#include "main.h"
#include "base58.h"

constexpr size_t nCoinDBCache = (1 << 31);
constexpr int maxOpenFiles = 1000;
constexpr int MAX_RECONSIDER = 100;
constexpr size_t MPOOL_SIZE = MAX_RECONSIDER * 4000000; // do not bother, we empty the mempool anyway

const char* usage = R"([-t] [-H height]

  -h, --help:    print this help message
  -t, --testnet: use testnet (default: mainnet)
  -H, --height:  use given height (default: best known tip)
)";

void print_usage(char const* self) {
    std::cout << "Usage: " << self << " " << usage << std::endl;
    exit(-1);
}

int main(int argc, char *argv[]) {
    bool testnet = false;
    int height = INT_MAX;

    char opt;
    int option_index = 0;
    static struct option long_options[] = {
        {"help",    no_argument,       0,  'h' },
        {"testnet", no_argument,       0,  't' },
        {"height",  required_argument, 0,  'H' },
        {0,         0,                 0,  0 }
    };

    while ((opt = getopt_long(argc, argv, "thH:", long_options, &option_index)) != -1) {
        switch (opt) {
            case 't':
                testnet = true;
                break;
            case 'h':
                print_usage(argv[0]);
                break;
            case 'H':
                height = atoi(optarg);
                if (height < 1) print_usage(argv[0]);
                break;
            case '?':
                print_usage(argv[0]);
                break;
      }
    }

    if (!testnet) {
        std::cerr << "Initializing MAINNET parameters" << std::endl;
        SelectParams(CBaseChainParams::MAIN);
    }
    else {
        std::cerr << "Initializing TESTNET parameters" << std::endl;
        SelectParams(CBaseChainParams::TESTNET);
    }

    const CChainParams& chainparams = Params();
    std::cerr << "Loading block index..." << std::flush;
    pblocktree = new CBlockTreeDB(nCoinDBCache, maxOpenFiles, false, false);
    CCoinsViewDB* pcoinsdbview = new CCoinsViewDB(nCoinDBCache, maxOpenFiles, false, false);
    pcoinsTip = new CCoinsViewCache(pcoinsdbview);
    LoadBlockIndex();
    std::cerr << " done!" << std::endl;

    int best_height = pcoinsTip->GetHeight();
    if (height != INT_MAX && (height <= best_height - MAX_RECONSIDER || height > best_height)) {
        std::cerr << "Invalid height requested; current height: " << best_height << std::endl;
        return -1;
    }

    uint256 inv_hash;
    CBlockIndex* pblockindex = nullptr;
    std::unique_ptr<ECCVerifyHandle> globalVerifyHandle(new ECCVerifyHandle());
    ECC_Start();
    mempool.reset(new CTxMemPool(CFeeRate(0), MPOOL_SIZE));
    if (height != INT_MAX && height < best_height) {
        std::cerr << "Setting desired height (" << height << ") ..." << std::flush;
        pblockindex = chainActive[height + 1];
        CValidationState state;
        InvalidateBlock(state, pblockindex);
        if (state.IsValid()) {
            ActivateBestChain(state);
        }
        mempool->clear();
        std::cerr << " done!" << std::endl;
    }
    FlushStateToDisk();

    size_t tot_known = 0;
    size_t tot_utxos = 0;
    size_t tot_amount = 0;
    size_t tot_p2pk, tot_p2pkh, tot_p2sh, tot_others;
    tot_p2pk = tot_p2pkh = tot_p2sh = tot_others = 0;

    std::cerr << "Dumping utxos..." << std::flush;
    std::unordered_map<std::string, AddressInfo> agg = pcoinsdbview->dumpUtxoSet();
    for (auto const& [addr, info]: agg) {
        switch (info.type) {
            case CScript::ScriptType::P2SH:
                tot_p2sh++;
                break;
            case CScript::ScriptType::P2PKH:
                tot_p2pkh++;
                break;
            case CScript::ScriptType::P2PK:
                tot_p2pk++;
                break;
            default:
                std::cerr << "Unknown address type found!" << std::endl;
                tot_others++;
        }

        std::cout << addr << "," << info.amount << "," << (int)info.type << std::endl;
        tot_utxos += info.count;
        tot_amount += info.amount;
    }
    std::cerr << " done!" << std::endl;
    std::cerr << "Tot addresses: " << agg.size() << std::endl;
    std::cerr << "Tot utxos: " << tot_utxos << std::endl;
    std::cerr << "Tot P2PK: " << tot_p2pk << std::endl;
    std::cerr << "Tot P2PKH: " << tot_p2pkh << std::endl;
    std::cerr << "Tot P2SH: " << tot_p2sh << std::endl;
    std::cerr << "Tot Others: " << tot_others << std::endl;

    std::cerr << "Checking correctness..." << std::flush;
    CCoinsStats stats;
    if (pcoinsdbview->GetStats(stats)) {
        assert(stats.nTransactionOutputs == tot_utxos);
        assert(stats.nTotalAmount == tot_amount);
    }
    std::cerr << " ok" << std::endl;

    if (pblockindex) {
        connman.reset(new CConnman()); // needed to prevent ActivateBestChain from crashing
        std::cerr << "Restoring previous height..." << std::flush;
        CValidationState state;
        ReconsiderBlock(state, pblockindex);
        if (state.IsValid()) {
            ActivateBestChain(state);
        }
        std::cerr << " ok" << std::endl;
    }
    FlushStateToDisk();
}
